Source/CBLRestReplicator+Backgrounding.m

```diff
@@ -65,39 +65,46 @@ - (void) endBackgrounding {
 
 // Called when the replicator goes idle (from -updateActive)
 - (void) okToEndBackgrounding {
-    dispatch_async(dispatch_get_main_queue(), ^{
-        if ([_bgMonitor hasBackgroundTask]) {
-            _deepBackground = YES;
-            [self updateSuspended];
-            LogTo(Sync, @"%@: Now idle; ending background task", self);
-            [_bgMonitor endBackgroundTask];  // will probably suspend the process immediately
-        }
-    });
+    if ([_bgMonitor hasBackgroundTask]) {
+        _deepBackground = YES;
+        [self updateSuspended];
+        LogTo(Sync, @"%@: Now idle; ending background task", self);
+        [_bgMonitor endBackgroundTask];  // will probably suspend the process immediately
+    }
+}
+
+- (void) updateSuspended {
+    BOOL suspended = (_filesystemUnavailable || _deepBackground);
+    self.suspended = suspended;
 }
 
 
 ////// All the methods below are called on the MAIN THREAD, not the replicator thread ////////
 
 
 - (void) appBackgrounding {
-    if (_active && [_bgMonitor beginBackgroundTaskNamed: self.description]) {
-        LogTo(Sync, @"%@: App backgrounding; starting temporary background task", self);
-    } else {
-        LogTo(Sync, @"%@: App backgrounding, but replication is inactive; suspending", self);
-        _deepBackground = YES;
-        [self updateSuspended];
-    }
+    [self.db doSync: ^{
+        if (_active && [_bgMonitor beginBackgroundTaskNamed: self.description]) {
```

Comment:
The doc-comment of -beginBackgroundTaskNamed says "Should be called on the main thread" (probably because it calls into `UIApplication`) but this is calling it from the db thread.

Looking at the git log of this file I see that commit 50cc50a from last November says "...Thread-safety improvements in CBLRestReplicator+Backgrounding: make sure UIApplication is only called on the main thread."

If you need to access `_active` on the db thread, can you do just that one check in the -doSync: block? Or more simply, make `_active` an atomic bool?

Reply:
making _active atomic doesn't work for the case of starting the replicator in the background. Checking _active could still come before the _active is actually set on the replicator thread when starting the replicator.

Reply:
FYI: UIApplication's beginbackgroundtaskwithname: and endBackgroundTask : can be safely called from non-main thread per https://developer.apple.com/documentation/uikit/uiapplication/1623051-beginbackgroundtaskwithname?language=objc.